sql: remove the planner ref from planNodes

knz opened the pull request:
The first commit from #20169.

Visible changes: the `*planner` reference is removed from planNodes.

Release notes: none.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/20154)
<!-- Reviewable:end -->

knz committed:
sql: remove the *planner ref from scanNode and deleteNode

knz committed:
sql: remove more planner references from controlJobNode and cancelQueryNode

knz committed:
sql: remove the planner reference from createViewNode

knz committed:
sql: remove the planner reference from createSequenceNode

knz committed:
sql: remove the planner reference from distinctNode

knz committed:
sql: remove the txn and distSQLPlanner references from explainDistSQLNode

knz committed:
sql: remove the planner reference from traceNode

knz committed:
sql: remove the planner reference from joinNode

knz committed:
sql: remove the planner reference from limitNode

knz committed:
sql: remove the planner reference from renderNode

knz committed:
sql: break the dependency applyLimit -> sortNode.p

knz committed:
sql: remove the planner reference from sortNode

knz commented:
cc @jordanlewis

knz committed:
sql: remove the planner reference from windowNode

knz commented:
@justinj I was advised by Vivek that you could help @jordanlewis with reviews around this area.

jordanlewis commented:
Haven't reviewed carefully, but blanket +1 on removing references to `planner` from plan nodes. I suggest limiting the scope of this PR to exactly that if you have not already.

knz commented:
ok they're all gone.

knz committed:
sql: remove the planner reference from groupNode

knz commented:
Rebased on top of #20169 so no file renames, for your reviewing pleasure.

justinj commented on pkg/sql/group.go:
```diff
@@ -484,7 +481,12 @@ func (n *groupNode) addIsNotNullFilter(where *filterNode, render *renderNode) {
-func (n *groupNode) desiredAggregateOrdering() sqlbase.ColumnOrdering {
+//
+// TODO(knz/radu): it's expensive to instantiate the aggregate
+// function here just o determine whether it's a min or max. We could
```
`o -> to`

justinj commented on pkg/sql/render.go:
```diff
@@ -295,14 +293,14 @@ func (p *planner) SelectClause(
 // initFrom initializes the table node, given the parsed select expression
-func (r *renderNode) initFrom(
-	ctx context.Context, parsed *tree.SelectClause, scanVisibility scanVisibility,
+func (p *planner) initFrom(
```
any particular reason this and other methods moved from the renderNode to the planner?

justinj reviewed (approved):
General idea LGTM, I have some other questions that I messaged you about, but definitely in favour of this decoupling!

knz commented:
There's an error in the last commit. Fixing now.

---

Review status: 0 of 35 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pkg/sql/group.go, line 486 at r16](https://reviewable.io:443/reviews/cockroachdb/cockroach/20154#-KzOcM0jrwLhpQeRxcPO-r16-486:-KzUD5aqN7FKLwCGKyui:bt2vr1b) ([raw file](https://github.com/cockroachdb/cockroach/blob/41e503b20401e26ae118f64b6f1514c34c1539ba/pkg/sql/group.go#L486)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

`o -> to`
</blockquote></details>

Fixed.

---

*[pkg/sql/render.go, line 296 at r16](https://reviewable.io:443/reviews/cockroachdb/cockroach/20154#-KzOcM0kxygUzEUfMhdT-r16-296:-KzUD9_wjJtfMoNCJoJM:b-4jwlc) ([raw file](https://github.com/cockroachdb/cockroach/blob/41e503b20401e26ae118f64b6f1514c34c1539ba/pkg/sql/render.go#L296)):*
<details><summary><i>Previously, justinj (Justin Jaffray) wrote…</i></summary><blockquote>

any particular reason this and other methods moved from the renderNode to the planner?
</blockquote></details>

This function calls `(*planner).getSources()` to construct the logical plan corresponding to the FROM clause. The `planner` reference has to come from somewhere. Si it's either making it a method of `*planner` or passing `p*planner`  as explicit argument. 

Now consider the name "planner". The planner object has too many responsibilities right now, but its name is clear: its purpose, the only purpose it *eventually* will fulfill, is to take an AST and construct a logical plan. So a method that takes a FROM AST and returns a logical plan (or, in this case, annotates it in a pre-initialized logical plan node) is the proper design.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20154#-:-KzUDnpRCsYnVUPgLGmq:b-1j6x2s)*
<!-- Sent from Reviewable.io -->

knz committed:
sql: make sortNode use its own struct as in-memory sorting container

The local execution logic was using the valuesNode (a logical plan
concept) as run-time row container (a strictly execution-time concept)
to perform the sort. This brought two unrelated concerns into the code
around `valuesNode`, making it harder to work with.

This patch moves towards a separation of concerns by adding a new
`sortValues` struct to do the in-memory sorting in local execution,
independent from valuesNode.

knz committed:
sql: remove the valuesNode code taken over by sortValues

knz committed:
sql: remove the planner reference from valuesNode

knz commented:
I had forgotten a step because of some oddity in the distsql physical planner. Added an intermediate commit that addresses the problem. The PR should be ready now. PTAL. 

---

Review status: 0 of 45 files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20154#-:-KzWExBV8K2hvAxtKb3p:bca5bfa)*
<!-- Sent from Reviewable.io -->

knz committed:
sql: plumb an evalContext in runParams for local execution

During distsql physical planning, even when local execution is not
used, the distsql physical planner calls `Start()` on valuesNodes, in
order to have it compute the values from the expression tuples. This
in turn needs to use a `tree.EvalContext` and allocates a row container
which, in turn, must register its memory usage against some memory
monitor.

Prior to this patch, the EvalContext and memory monitor used for this
reservation were those stored in the planner *whose reference was
stored in the valuesNode*. So we had the following chain of
responsibility:

- the planner has a reference to the Session which has a current "txn
  state" which has a memory monitor. The planner also has an
  EvalContext.

- during logical plan construction by the planner, the valuesNode
  would store a reference to the planner.

- during distsql physical planning, the distsql planner would ask the
  valuesNode to "compute itself", which would access the planner's
  EvalContext, and also allocate a rows container, which would access
  the previously stored planner reference, to access the session, to
  find a bytes monitor.

This is a bit too convoluted and instead we wish to be able to process
valuesNodes without direct access to a planner, because after all they
just need a valid EvalContext and valid monitor.

To achieve this, we start by observing that the EvalContext already
has a valid monitor (`EvalContext.Mon`). Then this patch changes
things as follows:

- the `runParams` struct, used by all planNodes' Start() method
  including `valueNode`'s, is extended with a pointer `evalCtx` to a
  `tree.EvalContext`.

- the various planNodes are modified to use that new
  `runParams.evalCtx` instead of `runParams.p.evalCtx` previously.

- the method implementation `(*valuesNode).Start()` is modified
  to use the memory monitor available from `runParams`, instead
  of the one provided by its planner reference.

- the distsql physical planner, is told to plumb its known valid
  EvalContext via `runParams`, even though it knows nothing about
  `planner`.

- every other instantiation of `runParams` is extended to populate
  the `evalCtx` field.

This patch incidentally breaks the dependency from the distsql
physical planner to the valueNode's planner reference. This was really
the primary motivation for this change.

knz committed:
sql: remove the planner reference from the subquery struct

knz committed:
sql: remove the Session reference from copyNode

RaduBerinde commented on pkg/sql/trace.go:
```diff
@@ -43,9 +42,9 @@ type traceNode struct {
-	// recording is set if this node started tracing on the session. If it did,
-	// then Close() needs to stop the recording.
-	recording bool
+	// stopTracing is set is set if this node started tracing on the
```
is set is set. Also "Close() must close it" is confusing

Why is this a function? It seems that the only purpose is so we can implicitly store `params.p.session` in it (so we're still storing runtime state, we just "hide" it a bit).

RaduBerinde commented:
LGTM

knz commented:
Review status: 0 of 46 files reviewed at latest revision, 3 unresolved discussions.

---

*[pkg/sql/trace.go, line 45 at r23](https://reviewable.io:443/reviews/cockroachdb/cockroach/20154#-KzOcM0mxRZAT9iAmvW7-r23-45:-KztgonGcXLv3MWgVzCS:b-aig6ka) ([raw file](https://github.com/cockroachdb/cockroach/blob/2ab209691caa92dd2dc4fdbdff6c6c5552a26fc5/pkg/sql/trace.go#L45)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

is set is set. Also "Close() must close it" is confusing

Why is this a function? It seems that the only purpose is so we can implicitly store `params.p.session` in it (so we're still storing runtime state, we just "hide" it a bit).
</blockquote></details>

It reduces the range of operations that can be performed on the stored session. This way, the only thing the planNode can do is stopping the tracing. Keeping the planner/session around would leave the door open to arbitrary other things.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20154)*
<!-- Sent from Reviewable.io -->

knz commented:
TFYRs
